### PR TITLE
Fixes #36978 - Add possibility to use remote webdriver

### PR DESCRIPTION
**Pull request by dosas**
In order to easily execute the selenium tests it would be good to have the possibility to use selenium remote webdriver images

```
docker run -d -p 4444:4444 --shm-size="2g" selenium/standalone-chrome:119.0
```

To access the driver via vagrant or a CI docker image the IP address needs to be set using an evironment variable:
```
SELENIUM_REMOTE_HOST=<ip_of_docker_host>
JS_TEST_DRIVER=selenium_chrome_remote
```

There are certain environments that require different chrome options to work so it is possible to add them via semicolon separated string as environment variable:
```
ADDITIONAL_CHROME_OPTIONS="--disable-dev-shm-usage"
```

### CODE REVIEW

**Comment by evgeni** [comment]
ok to test

**Comment by evgeni** [comment]
@MariaAga would you mind looking at this? Looks totally reasonable to me, but my knowledge around the integration tests is rather low.

File: test/integration_test_helper.rb [hunk 1]
```diff
@@ -20,18 +20,40 @@
 end
 
 Selenium::WebDriver::Chrome::Service.driver_path = ENV['TESTDRIVER_PATH'] || Foreman::Util.which('chromedriver', Rails.root.join('node_modules', '.bin'))
+
+options = Selenium::WebDriver::Chrome::Options.new
+options.args << '--disable-gpu'
+options.args << '--no-sandbox'
+options.args << '--window-size=1024,768'
```
Comment (stejskalleos) [inline test/integration_test_helper.rb]: Do we support this resolution, shouldn't it be 1920,1080?

Comment (evgeni) [reply test/integration_test_helper.rb]: That part just moved code.

So while "yeah, probably", it's not a change in this PR :)

File: test/integration_test_helper.rb [hunk 2]
```diff
@@ -20,18 +20,40 @@
 end
 
 Selenium::WebDriver::Chrome::Service.driver_path = ENV['TESTDRIVER_PATH'] || Foreman::Util.which('chromedriver', Rails.root.join('node_modules', '.bin'))
+
+options = Selenium::WebDriver::Chrome::Options.new
```
Comment (ekohl) [inline test/integration_test_helper.rb]: Why does it move options? I'd like to avoid defining variables in the global scope. Especially one so commonly named like `options`.

File: test/integration_test_helper.rb [hunk 3]
```diff
@@ -20,18 +20,40 @@
...
+
 Capybara.register_driver :selenium_chrome do |app|
-  options = Selenium::WebDriver::Chrome::Options.new
-  options.args << '--disable-gpu'
-  options.args << '--no-sandbox'
-  options.args << '--window-size=1024,768'
-  options.args << '--headless' unless ENV['DEBUG_JS_TEST'] == '1'
-  Capybara::Selenium::Driver.new(app, browser: :chrome, options: options)
+  if selenium_remote_host
+    Capybara::Selenium::Driver.new(
```
Comment (ekohl) [inline test/integration_test_helper.rb:39]: The `selenium_chrome` driver is a built in that has a defined meaning. See https://github.com/teamcapybara/capybara#selenium

I'd prefer to register this as a separate driver so you can use `JS_TEST_DRIVER=selenium_remote` instead.

I also question if the remote driver even understands the Chrome options.

File: test/integration_test_helper.rb [hunk 4]
```diff
@@ -20,18 +20,40 @@
...
 Selenium::WebDriver::Chrome::Service.driver_path = ENV['TESTDRIVER_PATH'] || Foreman::Util.which('chromedriver', Rails.root.join('node_modules', '.bin'))
+
+options = Selenium::WebDriver::Chrome::Options.new
+options.args << '--disable-gpu'
+options.args << '--no-sandbox'
+options.args << '--window-size=1024,768'
+options.args << '--headless' unless ENV['DEBUG_JS_TEST'] == '1'
+
+selenium_remote_host = ENV.fetch('SELENIUM_REMOTE_HOST', nil)
+selenium_remote_port = ENV.fetch('SELENIUM_REMOTE_PORT', 4444)
```
Comment (ekohl) [inline test/integration_test_helper.rb]: Can't this be done inside the registration bit instead of on the global scope?

File: test/test_helper.rb [hunk 5]
```diff
@@ -20,7 +20,11 @@
 FactoryBot.use_parent_strategy = false
 
 # Do not allow network connections and external processes
-WebMock.disable_net_connect!(allow_localhost: true)
+if ENV.fetch('SELENIUM_REMOTE_HOST', nil)
+  WebMock.allow_net_connect!
```
Comment (ekohl) [inline test/test_helper.rb]: Can we at least scope this more narrow to the remote host?
```suggestion
  WebMock.disable_net_connect!(allow_localhost: true, allow: ENV['SELENIUM_REMOTE_HOST'])
```

It may also need to allow the right port. See https://github.com/bblimke/webmock#external-requests-can-be-disabled-while-allowing-specific-requests

File: test/integration_test_helper.rb [hunk 6]
```diff
@@ -20,18 +20,40 @@
...
 Capybara.configure do |config|
-  config.javascript_driver      = ENV["JS_TEST_DRIVER"]&.to_sym || :selenium_chrome
-  config.default_max_wait_time  = 20
+  config.javascript_driver = ENV["JS_TEST_DRIVER"]&.to_sym || :selenium_chrome
+  config.default_max_wait_time = 20
   config.enable_aria_label = true
+  if selenium_remote_host
+    config.server_host = Socket.ip_address_list
+                           .find(&:ipv4_private?)
+                           .ip_address
```
Comment (ekohl) [inline test/integration_test_helper.rb]: What's the logic behind this? Isn't `::` (all IPs) a better choice?
```suggestion
    config.server_host = '::'
```

**Review by ekohl — CHANGES_REQUESTED** [review]
I've opened https://github.com/theforeman/foreman/pull/9962 which should provide a cleaner base for this PR.

File: test/integration_test_helper.rb (hunk 1 above)
Comment (MariaAga) [reply test/integration_test_helper.rb]: Also I think some tests relay on the screen to be that size.

Comment (ekohl) [reply test/integration_test_helper.rb]: In https://github.com/theforeman/foreman/pull/9962 I'm dropping it. That currently fails, but I don't know if it's because of me or something else.

File: test/integration_test_helper.rb (hunk 3 above)
Comment (dosas) [reply test/integration_test_helper.rb:39]: I will register a new symbol called `:selenium_chrome_remote`

The remote driver requires the options to be set:
```
msg: ArgumentError: Selenium::WebDriver::Remote::Driver needs :options to be set
```

But this can be solved by just passing the initialized object. Depending on the [discussion from above](https://github.com/theforeman/foreman/pull/9952#discussion_r1431767258) if all options can be removed.

File: test/integration_test_helper.rb (hunk 4 above)
Comment (dosas) [reply test/integration_test_helper.rb]: This can of course be done inside the registration. Currently the registration depends on `selenium_remote_host`.

Would you always register the `:selenium_chrome_remote` or would you make it dependent on `JS_TEST_DRIVER`?

I would go for always register so we can use the symbol for other comparisons:
i.e.
```
if ENV.fetch("JS_TEST_DRIVER", nil) == :selenium_chrome_remote.to_s
    config.server_host = "0.0.0.0"
```

File: test/integration_test_helper.rb (hunk 6 above)
Comment (dosas) [reply test/integration_test_helper.rb]: It looks like ipv6 is currently not possible with capybara or selenium:
```
 {"url":"http://:::41231/users/login"}
2023-12-20 08:47:01 INFO Selenium <- {"value":{"error":"invalid argument","message":"invalid argument\n  (Session info: headless chrome=119.0.6045.199)"
```
But we could try with `0.0.0.0`?

File: test/integration_test_helper.rb (hunk 1 above)
Comment (dosas) [reply test/integration_test_helper.rb]: @ekohl See my comment https://github.com/theforeman/foreman/pull/9962/files#r1432594792 to possible cause of error

**Comment by dosas** [comment]
I have no idea why the labeler fails maybe someone can help?

**Comment by evgeni** [comment]
> I have no idea why the labeler fails maybe someone can help?

that's just broken, ignore it for now :)

File: test/integration_test_helper.rb (hunk 1 above)
Comment (dosas) [reply test/integration_test_helper.rb]: @ekohl I based my PR on your changes from https://github.com/theforeman/foreman/pull/9962 and now the integration tests are failing.

I would assume this is because of the missing screen size option?

I would propose to first fix and merge your PR and then we can continue with this one.

File: test/integration_test_helper.rb (hunk 4 above)
Comment (ekohl) [reply test/integration_test_helper.rb]: I'd also always register it. Seems more straight forward

File: test/integration_test_helper.rb (hunk 6 above)
Comment (ekohl) [reply test/integration_test_helper.rb]: It may work with [::] but otherwise 0.0.0.0 probably works too. Fixing capybara to escape IPv6 would be a good thing

File: test/integration_test_helper.rb (hunk 1 above)
Comment (dosas) [reply test/integration_test_helper.rb]: Since there wasn't any progress on #9962 I incorporate all the changes here.

There is still one integration test failing. This might be due to the screen size option?

About the removed `--headless` option: The only way I found to pass chrome options to capybara is to register the driver. When doing this the options set in capybara https://github.com/teamcapybara/capybara/blob/3.39.2/lib/capybara/registrations/drivers.rb#L31 are overwritten. So if we want to set additional options like screen size we have to set `--headless` for `:selenium_chrome_headless`.  

Let me know in which direction we should move. Any help is greatly appreciated.

Comment (MariaAga) [reply test/integration_test_helper.rb]: The failing test is probably due to screen size. it tests if there is a column named "locations" in smart proxies page, but its hidden when the screen size is less than `991px`

File: test/integration_test_helper.rb [hunk 7]
```diff
@@ -20,18 +20,49 @@
...
   options = Selenium::WebDriver::Chrome::Options.new
-  options.args << '--disable-gpu'
-  options.args << '--no-sandbox'
-  options.args << '--window-size=1024,768'
-  options.args << '--headless' unless ENV['DEBUG_JS_TEST'] == '1'
-  Capybara::Selenium::Driver.new(app, browser: :chrome, options: options)
+  additional_options = ENV.fetch('ADDITIONAL_CHROME_OPTIONS', '')
+  additional_options.split(";").each do |add_opt|
+    options.args << add_opt
+  end
```
Comment (ekohl) [inline test/integration_test_helper.rb]: Isn't this equivalent?
```suggestion
  options.args += ENV.fetch('ADDITIONAL_CHROME_OPTIONS', '').split(';')
```

File: test/integration_test_helper.rb [hunk 8]
```diff
@@ -20,18 +20,49 @@
...
 end
+
 Capybara.configure do |config|
-  config.javascript_driver      = ENV["JS_TEST_DRIVER"]&.to_sym || :selenium_chrome
-  config.default_max_wait_time  = 20
+  config.javascript_driver = javascript_driver
+  config.default_max_wait_time = 20
   config.enable_aria_label = true
+  if ENV.fetch("JS_TEST_DRIVER", nil) == :selenium_chrome_remote.to_s
```
Comment (ekohl) [inline test/integration_test_helper.rb]:
```suggestion
  if ENV.fetch("JS_TEST_DRIVER", nil) == 'selenium_chrome_remote'
```

**Review by ekohl — COMMENTED** [review]
Some small nitpicks.

A bigger question here is how we're testing in GitHub Actions today and if this would actually be better. I don't know the current chromedriver & chrome are installed.

File: test/integration_test_helper.rb (hunk 8 above)
Comment (dosas) [reply test/integration_test_helper.rb]: IMHO it is clearer to derive the string from the symbol instead of kind of duplicating the definition.

File: test/integration_test_helper.rb (hunk 1 above)
Comment (dosas) [reply test/integration_test_helper.rb]: Thank you, tests are green now. So I guess we need to keep the screen size option then.

File: test/integration_test_helper.rb (hunk 8 above)
Comment (ekohl) [reply test/integration_test_helper.rb]: I disagree. While it's minimal, you'd adding redundancy. If it was a constant, I'd agree. Now there's just additional mental processing required when you read this that it does in fact ends up being a string.

File: test/integration_test_helper.rb [hunk 9]
```diff
@@ -299,7 +344,7 @@ def database_cleaner_strategy
   end
 
   def login_admin
-    visit('/users/login') if Capybara.current_driver == :selenium_chrome
+    visit('/users/login') if Capybara.current_driver.to_s.include? "chrome"
```
Comment (ekohl) [inline test/integration_test_helper.rb]: I wonder if this shouldn't be always done for any selenium browser instead of chrome.
```suggestion
    visit('/users/login') if Capybara.current_driver.to_s.include? "selenium"
```

File: test/integration_test_helper.rb (hunk 8 above)
Comment (dosas) [reply test/integration_test_helper.rb]: Okay then. I changed it.

File: test/integration_test_helper.rb [hunk 10]
```diff
@@ -20,18 +20,63 @@
...
+  config.javascript_driver = javascript_driver
+  config.default_max_wait_time = 20
   config.enable_aria_label = true
+  if ENV.fetch("JS_TEST_DRIVER", nil) == 'selenium_chrome_remote'
+    app_host = ENV.fetch('APP_SERVER_HOST') do
+      Socket.ip_address_list
+        .find(&:ipv4_private?)
+        .ip_address
+    end
+    app_port = ENV.fetch('APP_SERVER_PORT', "8080")
```
Comment (ekohl) [inline test/integration_test_helper.rb:73]: What made you choose 8080? I'm not aware of any config that defaults to this.

What I would prefer is to use the random port selection that is there by default. So instead:
```ruby
config.server_port = ENV['APP_SERVER_PORT'].to_i if ENV.key?('APP_SERVER_PORT')

config.app_host = "http://#{app_host}:#{config.server_port}"
```

But I don't know if this actually works because it's only in `Capybara::Server.new` that the port is finally determined:
https://github.com/teamcapybara/capybara/blob/52eaecea6d154b7d664b0032cd1cbcad4788fe65/lib/capybara/server.rb#L35-L37

File: test/integration_test_helper.rb [hunk 11]
```diff
@@ -20,18 +20,63 @@
...
+    config.server_port = app_port
+    # application server
+    config.server_host = "0.0.0.0"
+    # address used by selenium host to connect to application server
+    config.app_host = "http://#{app_host}:#{app_port}"
```
Comment (ekohl) [inline test/integration_test_helper.rb:78]: From reading the code I'd think this is redundant if you already set `server_host` correctly. In fact, it will even understand HTTP vs HTTPS then. It would also avoid the whole determining a port issue I described above.

At least, when I read:

https://github.com/teamcapybara/capybara/blob/52eaecea6d154b7d664b0032cd1cbcad4788fe65/lib/capybara/session.rb#L266

And `server_url` is defined by:

https://github.com/teamcapybara/capybara/blob/52eaecea6d154b7d664b0032cd1cbcad4788fe65/lib/capybara/session.rb#L837-L839

And `base_url` is defined by:

https://github.com/teamcapybara/capybara/blob/52eaecea6d154b7d664b0032cd1cbcad4788fe65/lib/capybara/server.rb#L91-L93

File: test/integration_test_helper.rb [hunk 12]
```diff
@@ -20,18 +20,63 @@
...
-  Capybara::Selenium::Driver.new(app, browser: :chrome, options: options)
+  options.args += ENV.fetch('ADDITIONAL_CHROME_OPTIONS', '').split(';')
+  options
+end
+
+if javascript_driver == :selenium_chrome_remote
+  ShowMeTheCookies.register_adapter(:selenium_chrome_remote, ShowMeTheCookies::SeleniumChrome)
+
+  Capybara.register_driver :selenium_chrome_remote do |app|
+    selenium_remote_host = ENV.fetch('SELENIUM_REMOTE_HOST', nil)
```
Comment (ekohl) [inline test/integration_test_helper.rb]: It's required so it's better to explicitly error out if this isn't given.
```suggestion
    selenium_remote_host = ENV.fetch('SELENIUM_REMOTE_HOST')
```

File: test/integration_test_helper.rb [hunk 13]
```diff
@@ -20,18 +20,63 @@
 end
 
 Selenium::WebDriver::Chrome::Service.driver_path = ENV['TESTDRIVER_PATH'] || Foreman::Util.which('chromedriver', Rails.root.join('node_modules', '.bin'))
-Capybara.register_driver :selenium_chrome do |app|
+
+javascript_driver = ENV.fetch("JS_TEST_DRIVER") { ENV['DEBUG_JS_TEST'] ? :selenium_chrome : :selenium_chrome_headless }.to_sym
+
+def get_chrome_options
```
Comment (ekohl) [inline test/integration_test_helper.rb]: Ruby convention is to avoid `get_` in method names
```suggestion
def chrome_options
```

**Review by ekohl — COMMENTED** [review]
This is just me reading the code. I hope to find some time to actually play around with this myself.

File: test/integration_test_helper.rb (hunk 9 above)
Comment (dosas) [reply test/integration_test_helper.rb]: Good point, adapted.

File: test/integration_test_helper.rb (hunk 12 above)
Comment (dosas) [reply test/integration_test_helper.rb]: Makes sense. Adapted.

File: test/integration_test_helper.rb (hunk 13 above)
Comment (dosas) [reply test/integration_test_helper.rb]: Done

File: test/integration_test_helper.rb (hunk 10 above)
Comment (dosas) [reply test/integration_test_helper.rb:73]: This sadly does not work because in the config stage `config.server_port` is empty/not set.

There is no real reason to make `APP_SERVER_PORT` configurable. I made it configurable since I did not want to have a hardcoded port that cannot be adapted.

File: test/integration_test_helper.rb (hunk 11 above)
Comment (dosas) [reply test/integration_test_helper.rb:78]: foreman_selenium drawio

I am also not entirely happy with this solution but it is the only way I could find to get it to work (without changes to capybara).

When app_host is not set capypara tries to access the IP set by `config.server_host` (`0.0.0.0`):
```
http://10.11.12.1:4444/wd/hub/session/a9f8f81dc07d3ccebc426e537e22d29a/url | {"url":"http://0.0.0.0:46703/users/login"}
```
But the docker container running outside of the vagrant box needs the IP of the vagrant box (see picture)

Sadly capybara does not allow to just change the IP but requires the workaround by `app_host` which includes scheme, ip, and port.

**Comment by dosas** [comment]
Any news on this?

**Comment by dosas** [comment]
@ekohl what do we need to do to get this forward?

**Comment by dosas** [comment]
ping

**Comment by dosas** [comment]
@ekohl Do you need any additional information on the answers from above?

File: test/integration/compute_profile_js_test.rb [hunk 14]
```diff
@@ -51,6 +51,7 @@ class ComputeProfileJSTest < IntegrationTestWithJavascript
   test "create compute profile" do
     visit compute_profiles_path()
     click_on("Create Compute Profile")
+    page.driver.browser.file_detector = nil if page.driver.browser.respond_to?(:file_detector=)
```
Comment (ekohl) [inline test/integration/compute_profile_js_test.rb]: What is the file_detector?

Comment (dosas) [reply test/integration/compute_profile_js_test.rb]: This is a workaround for a [bug in selenium](https://stackoverflow.com/questions/70441796/selenium-webdriver-for-aws-device-farm-error-when-sending-period-keystroke-t) which was fixed the same way in [foreman_puppet](https://projects.theforeman.org/issues/37251)

Comment (ekohl) [reply test/integration/compute_profile_js_test.rb]: That's nasty. Is there a bug in report in capybara we can link to?

How about we create a helper like `work_around_selenium_file_detector_bug` in `test/integration_test_helper.rb` that does this? Then other plugins can also easily use this and we can easily identify where it's needed.

Comment (dosas) [reply test/integration/compute_profile_js_test.rb]: > That's nasty. Is there a bug in report in capybara we can link to?

I couldn't find one. Though from the so post I'm not sure if it is a bug/feature in selenium or capybara.


> How about we create a helper like `work_around_selenium_file_detector_bug` in `test/
integration_test_helper.rb` that does this? Then other plugins can also easily use this and we can easily identify where it's needed.

Makes sense.

File: test/integration_test_helper.rb [hunk 15]
```diff
@@ -47,6 +92,11 @@ class << self
   # Stop ActiveRecord from wrapping tests in transactions
   self.use_transactional_tests = false
 
+  def work_around_selenium_file_detector_bug
+    # see: https://stackoverflow.com/questions/70441796/selenium-webdriver-for-aws-device-farm-error-when-sending-period-keystroke-t
```
Comment (ekohl) [inline test/integration_test_helper.rb]: If I'm really pedantic I'd suggest yardoc tags (which we don't use enough in our codebase).
```suggestion
    # @see https://stackoverflow.com/questions/70441796/selenium-webdriver-for-aws-device-farm-error-when-sending-period-keystroke-t
    def work_around_selenium_file_detector_bug
```

File: test/integration_test_helper.rb [hunk 16]
```diff
@@ -20,18 +20,63 @@
...
+  end
+elsif javascript_driver == :selenium_chrome_headless
+  options = chrome_options
+  options.args << '--headless'
+  Capybara.register_driver :selenium_chrome_headless do |app|
+    Capybara::Selenium::Driver.new(
+      app,
+      browser: :chrome,
+      options: options)
+  end
```
Comment (ekohl) [inline test/integration_test_helper.rb:53]: Why is this needed? This should be available by default:
https://github.com/teamcapybara/capybara/blob/0480f90168a40780d1398c75031a255c1819dce8/lib/capybara/registrations/drivers.rb#L31-L42

I had hoped to utilize this if possible. Given we don't set the window size elsewhere, I'd be tempted to drop `chrome_options`. I'm not sure anyone actually uses `ADDITIONAL_CHROME_OPTIONS` anyway.

File: test/integration_test_helper.rb [hunk 17]
```diff
@@ -20,18 +20,63 @@
...
+    selenium_remote_port = ENV.fetch('SELENIUM_REMOTE_PORT', 4444)
+    Capybara::Selenium::Driver.new(
+      app,
+      browser: :remote,
+      url: "http://#{selenium_remote_host}:#{selenium_remote_port}/wd/hub",
+      options: chrome_options)
+  end
+elsif javascript_driver == :selenium_chrome_headless
+  options = chrome_options
+  options.args << '--headless'
```
Comment (ekohl) [inline test/integration_test_helper.rb:47]: If we keep this, there's also a dedicated API:
```suggestion
  options.headless!
```

**Review by ekohl — DISMISSED** [review]

File: test/integration_test_helper.rb (hunk 15 above)
Comment (dosas) [reply test/integration_test_helper.rb]: I will change that but shouldn't we add a [rubocop](https://rubygems.org/gems/rubocop-yard/versions/0.9.3) for it then and if the community decides that this is a valid cop we can do it like that?

File: test/integration_test_helper.rb (hunk 17 above)
Comment (dosas) [reply test/integration_test_helper.rb:47]: I don't understand, can you please explain in more detail?

File: test/integration_test_helper.rb (hunk 16 above)
Comment (dosas) [reply test/integration_test_helper.rb:53]: As far as I remember (I touched this code 5 months ago) but the window size was required. Although the discussion if that is needed is IMO out of scope for this PR (can be done in follow up PR)

I'm requiring `ADDITIONAL_CHROME_OPTIONS` downstream. This will make it more easily configurable.

Comment (sbernhard) [reply test/integration_test_helper.rb:53]: I think its a good idea to be able to set additonal options which can be used. They are optional. So, if you don't need them, don't set the option and it doesn't hurt.

**Comment by sbernhard** [comment]
Looks like, this does not work @dosas . Use the previous implementation?

`
PG::Coder.new(hash) is deprecated. Please use keyword arguments instead! Called from /home/runner/work/foreman/foreman/vendor/bundle/ruby/3.0.0/gems/activerecord-6.1.7.8/lib/active_record/connection_adapters/postgresql_adapter.rb:883:in `new'
rake aborted!
NoMethodError: undefined method `headless!' for #<Selenium::WebDriver::Chrome::Options:0x000055b424d96268 @options={:args=>["--window-size=1024,768"], :prefs=>{}, :emulation=>{}, :local_state=>{}, :exclude_switches=>[], :perf_logging_prefs=>{}, :window_types=>[], :browser_name=>"chrome"}, @profile=nil, @logging_prefs={}, @encoded_extensions=[], @extensions=[]>

`

File: test/integration_test_helper.rb (hunk 17 above)
Comment (dosas) [reply test/integration_test_helper.rb:47]: see https://github.com/theforeman/foreman/pull/9952#issuecomment-2230534786

**Review by sbernhard — APPROVED** [review]

**Comment by sbernhard** [comment]
Failed tests are not related.
